### PR TITLE
Create AppCatalog CRs for compatibility 

### PR DESCRIPTION
**Pull request by tomahawk28**
No issue yet but we briefly talked about how to create appcatalog CRs for compatibility with catalog CRs and we decided to create `appcatalogsync` resource as our solution.

https://gigantic.slack.com/archives/C76JX6YLQ/p1624451154111100
## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/controller/catalog/resource/appcatalogsync/create.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		// Return early. Only unique instance manages appcatalogentry CRs.
+		return nil
+	}
+
+	cr, err := key.ToCatalog(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if namespace := cr.GetNamespace(); namespace == metav1.NamespaceDefault || namespace == "giantswarm" {
```
Comment (tomahawk28) [inline service/controller/catalog/resource/appcatalogsync/create.go:25]: I think customers should not care about catalog CRs in the default namespace or giantswarm namespace.  So I would skip creating appcatalog CRs in these cases. WDYT?

File: service/controller/catalog/resource/appcatalogsync/create.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		// Return early. No need to reconcile catalog CRs in default namespace or giantswarm namespace.
+		return nil
+	}
+
+	appCatalogCR := &v1alpha1.AppCatalog{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        cr.GetName(),
```
Comment (tomahawk28) [inline service/controller/catalog/resource/appcatalogsync/create.go:32]: Copy over the same name, the same labels, and the same annotation.

File: service/controller/catalog/resource/appcatalogsync/delete.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if namespace := cr.GetNamespace(); namespace == metav1.NamespaceDefault || namespace == "giantswarm" {
+		// Return early. No need to reconcile catalog CRs in default namespace or giantswarm namespace.
+		return nil
+	}
+
+	err = r.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogs().Delete(ctx, cr.GetName(), metav1.DeleteOptions{})
```
Comment (tomahawk28) [inline service/controller/catalog/resource/appcatalogsync/delete.go:29]: I think we should clean up appCatalog CRs created by the app-operator.

File: service/controller/catalog/resource/appcatalogsync/create.go (hunk 1 above)
Comment (rossf7) [reply service/controller/catalog/resource/appcatalogsync/create.go:25]: Yes, I think that makes sense.

File: service/controller/catalog/resource/appcatalogsync/create.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		Spec: v1alpha1.AppCatalogSpec{
+			Description: cr.Spec.Description,
+			Title:       key.CatalogTitle(cr),
+			Storage: v1alpha1.AppCatalogSpecStorage{
+				Type: "helm",
+				URL:  key.CatalogStorageURL(cr),
+			},
+		},
+	}
+	_, err = r.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogs().Create(ctx, appCatalogCR, metav1.CreateOptions{})
```
Comment (rossf7) [inline service/controller/catalog/resource/appcatalogsync/create.go:45]: I think we also need to handle updates. As I can see the situation where the catalog URL is wrong when the CR is first created. WDYT?

File: service/controller/catalog/resource/appcatalogsync/create.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	if apierrors.IsAlreadyExists(err) {
+		// no-op
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.Debugf(ctx, "created appCatalog %#q for compatibility", cr.GetName())
```
Comment (rossf7) [inline service/controller/catalog/resource/appcatalogsync/create.go:53]: Can you add the usual logging with creating and created? Just for consistency.

**Review by rossf7 — COMMENTED** [review]
Looks good. Just a couple of things.

File: service/controller/catalog/resource/appcatalogsync/create.go (hunk 4 above)
Comment (tomahawk28) [reply service/controller/catalog/resource/appcatalogsync/create.go:45]: You're right, I will make it handle updates as well.

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,6 +9,7 @@ and this project's packages adheres to [Semantic Versioning](http://semver.org/s
 
 ### Changed
 
+- Create `AppCatalog` CRs out of `Catalog` CRs in the cluster namespace for compatibility.
```
Comment (rossf7) [inline CHANGELOG.md:12]:
```suggestion
- Create `AppCatalog` CRs from `Catalog` CRs for compatibility with existing app-operator releases.
```

**Review by rossf7 — APPROVED** [review]
LGTM

**Pull request merged**